lib/dom.js

```diff
@@ -4,6 +4,117 @@
+}
+
+type CryptoKey$Type =
+  | 'secret'
+  | 'public'
+  | 'private';
+type CryptoKey$Usages =
+  | 'encrypt'
+  | 'decrypt'
+  | 'sign'
+  | 'verify'
+  | 'deriveKey'
+  | 'deriveBits'
+  | 'wrapKey'
+  | 'unwrapKey';
+declare interface CryptoKey {
+    algorithm: Object; // readonly
+    extractable: Boolean; // readonly
+    type: CryptoKey$Type; // readonly
+    usages: CryptoKey$Usages[]; // readonly
+}
+
+// TODO: CryptoKeyPair
+declare interface CryptoKeyPair {
+
+}
+
+// TODO: JSONWebKey
+type JSONWebKey = {};
+
+type RandomSource =  {
+  getRandomValues(typedArray: $TypedArray): void;
+};
+
+type SubtleCrypto$Algorithm =
+  | {name: 'AES-CBC', iv: BufferSource}
+  | {name: 'AES-CTR', counter: BufferSource, length: number}
+  | {name: 'AES-GCM', iv: BufferSource, additionalData?: BufferSource, tagLength?: number}
+  | {name: 'RSA-OAEP', label?: string};
+type SubtleCrypto$DerivedKeyAlgorithm = {
+  name: DOMString,
+  length?: 128 | 192 | 256,
+};
+// TODO: fix types for parameters
+type SubtleCrypto$DeriveKeyAlgorithm =
+  | {name: 'ECDH', public: any}
+  | {name: 'DH', public: any}
+  | {name: 'PBKDF2', salt: BufferSource, iterations: number, hash: DOMString}
+  | {name: 'HKDF-CTR', hash: any, label: any, context: any};
+type SubtleCrypto$GenerateKeyAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW'
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA'
+  | 'ECDH'
+  | 'DH';
+type SubtleCrypto$HashAlgo =
+  | 'SHA-1'
+  | 'SHA-256'
+  | 'SHA-384'
+  | 'SHA-512';
+type SubtleCrypto$KeyFormat =
+  | 'raw'
+  | 'pkcs8'
+  | 'spki'
+  | 'jwk';
+type SubtleCrypto$SignAlgo =
+  | 'HMAC'
+  | 'RSASSA-PKCS1-v1_5'
+  | 'ECDSA';
+type SubtleCrypto$WrapAlgo =
+  | 'AES-CBC'
+  | 'AES-CTR'
+  | 'AES-GCM'
+  | 'RSA-OAEP'
+  | 'AES-KW';
+declare interface SubtleCrypto {
+  decrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<string>;
+  // TODO: fix type for 'algorithm' param
+  deriveBits(algorithm: any, baseKey: CryptoKey, length: number): Promise<ArrayBuffer>;
+  deriveKey(algorithm: SubtleCrypto$DeriveKeyAlgorithm, masterKey: CryptoKey, derivedKeyAlgorithm: DOMString | SubtleCrypto$DerivedKeyAlgorithm, extractable: Boolean, keyUsages: CryptoKey$Usages[]): Promise<CryptoKey | CryptoKeyPair>;
+  digest(algo: SubtleCrypto$HashAlgo, buffer: BufferSource): Promise<string>;
+  encrypt(algorithm: SubtleCrypto$Algorithm, key: CryptoKey, data: BufferSource): Promise<ArrayBuffer>;
+  exportKey(format: SubtleCrypto$KeyFormat, key: CryptoKey): Promise<CryptoKey>;
+  generateKey(algo: SubtleCrypto$GenerateKeyAlgo, extractable: Boolean, keyUsages: CryptoKey$Usages[]): Promise<CryptoKey | CryptoKeyPair>;
+  importKey(format: SubtleCrypto$KeyFormat, keyData: ArrayBuffer | JSONWebKey, algo: SubtleCrypto$GenerateKeyAlgo, extractable: Boolean, usages: CryptoKey$Usages[]): Promise<CryptoKey>;
+  sign(algo: SubtleCrypto$SignAlgo, key: CryptoKey, text2sign: BufferSource): Promise<BufferSource>;
```

Comment:
How did you choose to return a `BufferSource`-implementing type rather than just an `ArrayBuffer`? both **MDN** and **w3.org** itself fail to specify the content of the promise, at the time of writing. Could it depend on the algo?(unlikely)

Reply:
Because the input (`text2sign `) is of type `BufferSource`, it seemed logical the `Promise` would be something of `BufferSource` as well

Reply:
Because the input (`text2sign `) is of type `BufferSource`, it seemed logical the `Promise` would be something of `BufferSource` as well